controller/support_bundle_controller.go

```diff
@@ -0,0 +1,707 @@
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace, serviceAccount string) *SupportBundleController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	c := &SupportBundleController{
+		baseController: newBaseController("longhorn-support-bundle", logger),
+		controllerID:   controllerID,
+
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+
+		ds: ds,
+
+		kubeClient: kubeClient,
+		httpClient: &http.Client{Timeout: 30 * time.Second},
+
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: "longhorn-support-bundle-controller"}),
+	}
+
+	ds.SupportBundleInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueue,
+		UpdateFunc: func(old, cur interface{}) { c.enqueue(cur) },
+		DeleteFunc: c.enqueue,
+	})
+	c.cacheSyncs = append(c.cacheSyncs, ds.OrphanInformer.HasSynced)
+
+	return c
+}
+
+func (c *SupportBundleController) enqueue(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *SupportBundleController) enqueueAfter(obj interface{}, delay time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	c.queue.AddAfter(key, delay)
+}
+
+func (c *SupportBundleController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("Start Longhorn Support Bundle controller")
+	defer c.logger.Info("Shutting down Longhorn Support Bundle controller")
+
+	if !cache.WaitForNamedCacheSync(c.name, stopCh, c.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *SupportBundleController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SupportBundleController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncSupportBundle(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *SupportBundleController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	log := c.logger.WithField("supportBundle", key)
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		log.WithError(err).Warn("Error syncing Longhorn SupportBundle")
+
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	log.WithError(err).Warn("Dropping Longhorn SupportBundle out of the queue")
```

Comment:
1. We are not allowing more than 1 SupportBundle generating.
1. When the SupportBundle starts, each state change will return the queue.
1. When the SupportBundle starts, all unexpected errors should mark it failed. 
1. This is to handle errors outside the SupportBundle phase.

Reply:
My question is which code will handle this `When the SupportBundle starts, all unexpected errors should mark it failed.`  after dropping the support bundle key from the queue?

Reply:
We will hit `utilruntime.HandleError(err)` when `err != nil` and the queue has exhausted the `maxRetries(3)`.

1 After the SupportBundle starts, when the phase encounters an error, we will return `c.updateSupportBundleRecord`. This updates the record and returns a ` nil` error. 
```
record.recordType = supportBundleRecordError
record.nextState = longhorn.SupportBundleStateError
```
1. At the defer handleStatusUpdate will update the SupportBundle to an `Error` state and error condition.
1. Returns queue because of `nil` error.
1. Controller picks up again because of the error state update.
1. Do nothing then drops the queue.

This means it will not hit `utilruntime.HandleError(err)`  after SupportBundle starts because of the returning of the `nil` error.